plugins/org.eclipse.elk.core.meta/src/org/eclipse/elk/core/meta/jvmmodel/MetaDataJvmModelInferrer.xtend

```diff
@@ -293,35 +266,158 @@ class MetaDataJvmModelInferrer extends AbstractModelInferrer {
     '''
     
     private def StringConcatenationClient registerLayoutAlgorithms(MdBundle bundle) '''
-        «FOR algorithm : bundle.members.filter(MdAlgorithm)»
-            registry.register(new «LayoutAlgorithmData»(
+        «FOR MdAlgorithm algorithm : bundle.members.filter(MdAlgorithm)»
+            new «algorithm.qualifiedTargetClass»().apply(registry);
+        «ENDFOR»
+    '''
+    
+    
+    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // ALGORITHM METADATA PROVIDER CLASS GENERATION
+
+    /**
+     * Generates code for the given algorithm'm metadata provider class.
+     * 
+     * @param algorithm
+     *            the algorithm to create one or more {@link JvmDeclaredType declared types} from.
+     * @param acceptor
+     *            each created {@link JvmDeclaredType type} without a container should be passed to the acceptor
+     *            in order to get attached to the current resource. The acceptor's
+     *            {@link IJvmDeclaredTypeAcceptor#accept(org.eclipse.xtext.common.types.JvmDeclaredType) accept(..)}
+     *            method takes the constructed empty type for the pre-indexing phase. This one is further
+     *            initialized in the indexing phase using passed closure.
+     * @param isPreIndexingPhase
+     *            whether the method is called in a pre-indexing phase, i.e.
+     *            when the global index is not yet fully updated. You must not
+     *            rely on linking using the index if isPreIndexingPhase is
+     *            <code>true</code>.
+     */
+    private def void inferAlgorithm(MdAlgorithm algorithm, IJvmDeclaredTypeAcceptor acceptor,
+        boolean isPreIndexingPhase) {
+            
+        // If the algorithm doesn't have a name yet, don't bother
+        if (algorithm.name === null) {
+            return
+        }
+        
+        // Create a class for the general property definitions to be used by clients as well as for the registration
+        // method that instantiates the ILayoutMetaDataProviders
+        acceptor.accept(algorithm.toClass(algorithm.qualifiedTargetClass)) [
+            // Implement the ILayoutMetaDataProvider interface and... BE DOCUMENTED!
+            superTypes += typeRef(ILayoutMetaDataProvider)
+            fileHeader = algorithm.bundle.documentation
+            documentation = algorithm.documentation
+            
+            // 1. Public constants for supported layout properties
```

Comment:
Yup, I'll change that. I'll also take the opportunity to rename MdProperty to MdOption, as previously discussed.